allennlp_models/rc/dataset_readers/squad2.py

```diff
@@ -0,0 +1,197 @@
+import json
+import logging
+from typing import Any, Dict, List, Tuple, Optional
+
+from overrides import overrides
+
+from allennlp.common.file_utils import cached_path
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.instance import Instance
+from allennlp.data.token_indexers import SingleIdTokenIndexer, TokenIndexer
+from allennlp.data.tokenizers import Token, Tokenizer, SpacyTokenizer
+
+from allennlp_models.rc.dataset_readers import utils
+
+logger = logging.getLogger(__name__)
+SQUAD2_NO_ANSWER_TOKEN = "@@<NO_ANSWER>@@"
+
+
+@DatasetReader.register("squad2")
+class Squad2Reader(DatasetReader):
+    """
+    Reads a JSON-formatted SQuAD 2.0 file and returns a ``Dataset`` where the ``Instances`` have four
+    fields: ``question``, a ``TextField``, ``passage``, another ``TextField``, and ``span_start``
+    and ``span_end``, both ``IndexFields`` into the ``passage`` ``TextField``.  We also add a
+    ``MetadataField`` that stores the instance's ID, the original passage text, gold answer strings,
+    and token offsets into the original passage, accessible as ``metadata['id']``,
+    ``metadata['original_passage']``, ``metadata['answer_texts']`` and
+    ``metadata['token_offsets']``. This is so that we can more easily use the official SQuAD 2.0
+    evaluation script to get metrics.
+
+    This differs from the SquadReader in that we append a special no-answer token to each
+    context. If the model's argmax span is this token, then we predict that the question is unanswerable,
+    given the provided context.
```

Comment:
I haven't read the rest yet, but if this is the only difference, can't we make it a flag on the original reader?

Reply:
Yeah, that would work too!